pkgs/development/beam-modules/default.nix

```diff
@@ -1,16 +1,53 @@
-{ stdenv, pkgs }:
+{ stdenv, pkgs, erlang, overrides ? (self: super: {}) }:
 
 let
-  self = rec {
-    hexPackages = import ./hex-packages.nix { stdenv = stdenv; callPackage = self.callPackage; pkgs = pkgs; };
-    callPackage = pkgs.lib.callPackageWith (pkgs // self // hexPackages);
-    buildRebar3 = callPackage ./build-rebar3.nix {};
-    buildHex = callPackage ./build-hex.nix {};
-    buildErlangMk = callPackage ./build-erlang-mk.nix {};
-    buildMix = callPackage ./build-mix.nix {};
-
-    ## Non hex packages
-    hex = callPackage ./hex {};
-    webdriver = callPackage ./webdriver {};
-  };
-in self // self.hexPackages
+  inherit (stdenv.lib) fix' extends getVersion versionAtLeast;
+
+  lib = pkgs.callPackage ./lib.nix {};
+
+  # FIXME: add support for overrideScope
+  callPackageWithScope = scope: drv: args: stdenv.lib.callPackageWith scope drv args;
+  mkScope = scope: pkgs // scope;
+
+  packages = self:
+    let
+      defaultScope = mkScope self;
+      callPackage = drv: args: callPackageWithScope defaultScope drv args;
+    in
+      import ./hex-packages.nix {
+        inherit pkgs stdenv callPackage;
+      } // {
+        inherit callPackage erlang;
+        beamPackages = self;
+
+        rebar = callPackage ../tools/build-managers/rebar { };
+        rebar3-open = callPackage ../tools/build-managers/rebar3 {
+          hermeticRebar3 = false;
+        };
+        rebar3 = callPackage ../tools/build-managers/rebar3 {
+          hermeticRebar3 = true;
+        };
+
+        hexRegistrySnapshot = callPackage ./hex-registry-snapshot.nix { };
+        fetchHex = callPackage ./fetch-hex.nix { };
+
+        buildRebar3 = callPackage ./build-rebar3.nix {};
+        buildHex = callPackage ./build-hex.nix {};
+        buildErlangMk = callPackage ./build-erlang-mk.nix {};
+        buildMix = callPackage ./build-mix.nix {};
+
+        # BEAM-based languages.
+        elixir = if versionAtLeast (lib.getVersion erlang) "18"
+                 then callPackage ../interpreters/elixir { debugInfo = true; }
+                 else throw "Elixir requires at least Erlang/OTP R18.";
+        lfe = callPackage ../interpreters/lfe { };
+
+        # Non hex packages
+        hex = callPackage ./hex {};
+        webdriver = callPackage ./webdriver {};
+
+        hex2nix = callPackage ../tools/erlang/hex2nix {};
+        cuter = callPackage ../tools/erlang/cuter {};
+        relxExe = callPackage ../tools/erlang/relx-exe {};
+      };
+in fix' (extends overrides packages)
```

Comment:
@gleber For the sake of consistency, you might want to leverage [`makeExtensible`](https://github.com/NixOS/nixpkgs/blob/87b4a91fc4145e04ca028c08754bd144b8d1d02d/lib/fixed-points.nix#L70) here. It's presently used by both the Haskell package sets and `linuxPackages` sets.

The only change you'd need to make is replace this line with:

```
stdenv.lib.makeExtensible (extends overrides packages)
```

You could then remove the `overrides` argument, as you'll then be able to do `beam.packages.erlang.extend (self: super: { /* ... */ })`

Aside from aesthetics, I think there's a growing sentiment that "override" is a little bit overloaded, as it has historically been used to mean two different things:

* overriding the inputs into an expression, and
* overriding the packages within a package set

The use of "overriding" in the latter isn't exactly accurate, anyway -- you often want to "extend" the package set (and overriding can be seen as a subset of that concern). So the hope is that consistently using `.override` for modifying inputs and `.extend` for inputs will make things more intuitive, and having one way of going about it (`lib.makeExtensible`) will ensure that each package set gets it right, for free.

But, I'm not much involved in BEAM in Nixpkgs, so I figured I'd make the suggestion and leave the choice at your discretion :).